File: c++/67_Add_Binary.cpp
#include <iostream>
#include <vector>
#include <map>
#include <string>
#include <algorithm>

using namespace std;

class Solution {
public:
    string addBinary(string a, string b) {
    	int m = a.size();
    	int n = b.size();
    	if (m < n)
    		return addBinary(b, a);
    	int total, carry = 0;
    	string res;
    	for (int i = 0; i < m; i++) {
    		total = a[m - 1 - i] - '0' + carry;
    		if (i < n)
    			total += b[n - 1 - i] - '0';
    		cout << total << endl;
    		carry = total / 2;
    		res.push_back(total % 2 + '0');
    	}
    	if (carry)
    		res.push_back('1');
    	reverse(res.begin(), res.end());
    	return res;
    }
};


int main() {
	Solution s;
	cout << s.addBinary("11", "1") << endl;


}

File: python/67_Add_Binary.py
#!/usr/bin/python
import sys
import os

'''
a = "11"
b = "1"
Return "100".
'''

class Solution(object):
    def addBinary(self, a, b):
        if (len(a) < len(b)):
        	return self.addBinary(b, a)
        
        m, n = len(a), len(b)
        carry = 0
        res = []
        for i in range(m):
        	if i > n - 1:
        		total = int(a[m - 1 -i]) + int(carry)
        	else:
        		total = int(a[m - 1 -i]) + int(b[n - 1 -i]) + carry
        	carry = total / 2
        	#print i, a[i], b[], total, total % 2
        	res.append(str(total % 2))
        res = "".join(res)
        res = res[::-1]
        if carry:
        	res = "1" + res
        return res
        

def main():
    s = Solution()
    print s.addBinary("1010", "1011")

if __name__ == "__main__":
    main()
